subsys/net/l2/ethernet/arp.c

```diff
@@ -304,15 +311,30 @@ static inline struct net_pkt *arp_prepare(struct net_if *iface,
 
 	memset(&hdr->dst_hwaddr.addr, 0x00, sizeof(struct net_eth_addr));
 
-	net_ipaddr_copy(&hdr->dst_ipaddr, next_addr);
+	if (is_ipv4_auto && IS_ENABLED(CONFIG_NET_IPV4_AUTO)) {
+#if defined(CONFIG_NET_IPV4_AUTO)
+		net_ipaddr_copy(&hdr->dst_ipaddr,
+				&net_if_get_config(iface)->
+				ipv4auto.requested_ip);
+#endif
+	} else {
+		net_ipaddr_copy(&hdr->dst_ipaddr, next_addr);
+	}
 
 	memcpy(hdr->src_hwaddr.addr, eth->src.addr,
 	       sizeof(struct net_eth_addr));
 
 	if (entry) {
 		my_addr = if_get_addr(entry->iface);
 	} else {
-		my_addr = &NET_IPV4_HDR(pending)->src;
+		if (is_ipv4_auto && IS_ENABLED(CONFIG_NET_IPV4_AUTO)) {
+#if defined(CONFIG_NET_IPV4_AUTO)
```

Comment:
This is awful: IS_ENABLED _and_ #ifdef used side by side? 
To avoid that, make inline function returning pointer on ipv4auto address in net_if, if not enabled: just make stub ones that return NULL.